container_api/src/container/ui_admin.rs

```diff
@@ -0,0 +1,389 @@
+        self.save_config()?;
+
+        self.static_servers
+            .remove(id)
+            .ok_or(HolochainError::ErrorGeneric(
+                "Could not remove server".into(),
+            ))?;
+
+        Ok(())
+    }
+
+    fn start_ui_interface(&mut self, id: &String) -> Result<(), HolochainInstanceError> {
+        let server = self.static_servers.get_mut(id)?;
+        notify(format!("Starting UI interface \"{}\"...", id));
+        server.start()
+    }
+
+    fn stop_ui_interface(&mut self, id: &String) -> Result<(), HolochainInstanceError> {
+        let server = self.static_servers.get_mut(id)?;
+        notify(format!("Stopping UI interface \"{}\"...", id));
+        server.stop()
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use container::admin::tests::*;
+    use std::{fs::File, io::Read};
+
+    #[test]
+    fn test_install_ui_bundle_from_file() {
+        let mut container = create_test_container("test_install_ui_bundle_from_file", 3000);
+        let bundle_path = PathBuf::from(".");
+        assert_eq!(
+            container.install_ui_bundle_from_file(bundle_path, &"test-bundle-id".to_string()),
+            Ok(())
+        );
+    }
+
+    #[test]
+    fn test_uninstall_ui_bundle() {
+        let mut container = create_test_container("test_uninstall_ui_bundle", 3000);
+        assert_eq!(
+            container.uninstall_ui_bundle(&"test-bundle-id".to_string()),
+            Err(HolochainError::ConfigError(
+                "No UI bundles match the given ID \"test-bundle-id\"".into()
+            ))
+        );
+        let bundle_path = PathBuf::from(".");
+        assert_eq!(
+            container.install_ui_bundle_from_file(bundle_path, &"test-bundle-id".to_string()),
+            Ok(())
+        );
+        assert_eq!(
+            container.uninstall_ui_bundle(&"test-bundle-id".to_string()),
+            Ok(())
+        );
+        let mut config_contents = String::new();
+        let mut file = File::open(&container.config_path).expect("Could not open temp config file");
+        file.read_to_string(&mut config_contents)
+            .expect("Could not read temp config file");
+
+        let mut toml = String::from("bridges = []\nui_bundles = []\nui_interfaces = []");
+        toml = add_block(toml, agent1());
+        toml = add_block(toml, agent2());
+        toml = add_block(toml, dna());
+        toml = add_block(toml, instance1());
+        toml = add_block(toml, instance2());
+        toml = add_block(toml, interface(3000));
+        toml = add_block(toml, logger());
+        toml = format!("{}\n", toml);
+
+        assert_eq!(config_contents, toml,)
+    }
+
+    #[test]
+    fn test_add_ui_interface() {
+        let mut container = create_test_container("test_add_ui_interface", 3000);
+        assert_eq!(
+            container.add_ui_interface(UiInterfaceConfiguration {
+                id: "test-ui-interface-id".into(),
+                port: 3000,
```

Comment:
I think you need to use a different port number for each test, because on some platforms (travis/windows) they run asynchronously and we get a port in-use error.

Reply:
Yeah, and in this case, you are using the same port for the websocket and the UI interface in the same test (3000 in line 207 above). This won't work.

Reply:
Great to know to keep this in mind. All tests now use different ports for both UI static servers and interfaces